draft/ZEP0010.md

```diff
@@ -0,0 +1,413 @@
+extension is supported. If an extension includes ``"must_understand": true``
+(the default) and the implementation does not support it, the dataset must not
+be loaded and an appropriate error should be raised. For extensions with
+``"must_understand": false``, implementers may safely ignore unrecognized entries.
+
+To support a given extension, an implementation must either (1) check for known
+extension names and invoke appropriate logic according to the extension’s
+specification at the correct point in its processing pipeline (e.g., during
+metadata interpretation, data access, or layout resolution), or (2) delegate
+that logic via a callback or plugin mechanism that allows third-party code to
+handle the extension dynamically. This modular approach enables implementers to
+support a flexible and evolving set of extensions while maintaining core
+compatibility.
+
+### Examples
+
+The following examples represent a few realistic use cases of the top-level
+``extensions`` container. This ZEP is putting in place the mechanism so the
+community can experiment with such extensions before their standardization.
+
+#### Offset (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.offset",
+            "configuration": { "offset": [ 12, 24 ] }
+        }
+    ]
+}
+```
+
+The ``example.offset`` extension contains an array of the same order as the
+shape of the containing array specifying which element of the array should be
+considered as the origin, e.g., `[0, 0]`. This allows the reuse of subregions
+of an array without the need to rewrite the data.
+
+Note that in this example of the extension is ``must_understand=true`` meaning
+an implementation which does not support the ``example.offset`` extension
+should raise an error.
+
+#### Statistics (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.array-statistics",
+            "must_understand": false,
+            "configuration": {
+                "min": 5,
+                "max": 1023
+            }
+        }
+    ]
+}
+```
+
+The ``example.array-statistics`` extension contains two fields -- ``min``
+and ``max`` specifying the range of values which are present in the array,
+reducing the need to read every byte. ``must_understand`` is false, so
+implementations can safely ignored the extension.
+
+#### Skip empty chunks (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    ...,
+    "extensions": [
+        "example.skip_empty_chunks"
+    ]
+}
+```
+
+Currently the "write_empty_chunks" flag in zarr-python is not propagated
+to the zarr.json file. An extension like ``example.skip_empty_chunks``
+could serve as a no-configuration flag in the metadata to inform
+implementations that empty chunks should not be written.
+
+
+
+#### Domain metadata (group)
```

Comment:
This example goes against my own intuition regarding the extension vs. attribute distinction.  I assumed that extensions would be for things that require integration into the zarr implementation itself (i.e. zarr-python, tensorstore, zarrs) while attributes are for things that are handled by higher level layers, either "user" code or some higher-level abstraction like ome-zarr built on top of zarr.

Using extensions essentially just as namespaced attributes means that either:
- each implementation needs to have an implementation of the extension just to allow access to the attributes; or
- implementations need to allow users to directly read and write `must_understand: false` extension metadata.

Reply:
I updated the example, @jbms, but didn't fully address your question. I do see code extensions, e.g., ome-zarr having access to the extension metadata as one of three general options:
 - in implementation
 - in plugin (i.e. internal entrypoint)
 - in wrapper (i.e. higher-level)